File: iBeacon/JMCBeaconManager.m
//
//  JMCBeaconManager.m
//  iBeaconTest
//
//  Created by sadmin on 2/21/14.
//  Copyright (c) 2014 JanuszChudzynski. All rights reserved.
//
#import "JMCBeaconManager.h"


@import AudioToolbox;
@import CoreBluetooth;

@interface JMCBeaconManager()<CLLocationManagerDelegate, CBCentralManagerDelegate>
{
    CLProximity proximity;
    int counter;
    
}
@property(nonatomic,strong)CLLocationManager * locationManager;
@property(nonatomic,strong) CLBeacon * currentBeacon;
@property(nonatomic,strong) NSMutableArray * regions;
@property(nonatomic,strong) NSMutableDictionary * stateDictionary;
@property(nonatomic,strong) CBCentralManager * bluetoothManager;
@property BOOL bluetoothEnabled;
@end


@implementation JMCBeaconManager

/**Sometimes ios doesn't call did enter region in this case we will call it manually*/
-(void)updateState:(CLBeaconRegion *)region state:(CLRegionState)state{
    if([[_stateDictionary objectForKey:region.identifier]integerValue]!=state ){
        [_stateDictionary setObject:@(state) forKey:region.identifier];
    }
}

/***/
-(void)checkStateForRegion:(CLBeaconRegion *)region{
    if([[_stateDictionary objectForKey:region.identifier]integerValue]!=CLRegionStateInside){
       [self.locationManager requestStateForRegion:region];
    }
}


- (void)startBluetoothStatusMonitoring {

    self.bluetoothManager = [[CBCentralManager alloc]
                             initWithDelegate:self
                             queue:dispatch_get_main_queue()
                             options:@{CBCentralManagerOptionShowPowerAlertKey: @(NO)}];
}

#pragma mark - CBCentralManagerDelegate

- (void)centralManagerDidUpdateState:(CBCentralManager *)central {
    
    if ([central state] == CBCentralManagerStatePoweredOn) {
        self.bluetoothEnabled = YES;
    }
    else {
        self.bluetoothEnabled = NO;
        [self logMessage:@"You must enable bluetooth!"];
        
    }
}



-(void)logMessage:(NSString *)message{
    
    message =[NSString stringWithFormat:@"%@\r\n %@ \r\n %@", [NSDate new],message,self.logView.text];
    self.logView.text = message;

    NSLog(@"\n %@ \n ",message);
   // [self saveLog:message];
}


-(id)init{
    self = [super init];
    if(self){
        _locationManager = [[CLLocationManager alloc]init];
        
      //  [_locationManager requestWhenInUseAuthorization];
        _regions = [NSMutableArray new];
        _locationManager.delegate = self;
              counter =0;
        _stateDictionary = [NSMutableDictionary new];
        
    }
    return self;
}

/**
    Checks if iBeacon monitoring is supported
 */
-(BOOL)isSupported:(NSMutableString*)message{
    BOOL enabled = NO;
    [self startBluetoothStatusMonitoring];

    NSMutableString * msg = [NSMutableString new];
    [msg appendFormat:@""];
    if([CLLocationManager isMonitoringAvailableForClass:[CLRegion class]]){
        enabled = YES;
    }
    else{

        enabled = NO;
       [message appendFormat:@"%@ /n %@ ",message, @"Region Monitoring is not available on this device"];
    }
    

    
    if([CLLocationManager authorizationStatus ]== kCLAuthorizationStatusAuthorizedAlways ){
        enabled = YES && enabled;
    }
    else{
        enabled = NO;//&&enabled;
       [message appendFormat:@"%@ /n %@ ",message, @"Applications must be explicitly authorized to use location services by the user and location services must themselves currently be enabled for the system."];
    }


    if ([[UIApplication sharedApplication] backgroundRefreshStatus] == UIBackgroundRefreshStatusAvailable) {
        enabled = YES && enabled;

    }else if([[UIApplication sharedApplication] backgroundRefreshStatus] == UIBackgroundRefreshStatusDenied)
    {
      [message appendFormat:@"%@ /n %@ ",message, @"The user explicitly disabled background behavior for this app or for the whole system."];
        
        enabled = NO;// && enabled;
    }
    else if([[UIApplication sharedApplication] backgroundRefreshStatus] == UIBackgroundRefreshStatusRestricted)
    {
      
        [message appendFormat:@"%@ /n %@ ",message, @"unavailable on this system due to device configuration; the user cannot enable the feature."];
        
        enabled = NO;//&&enabled;

    }
    
    if(enabled) {
        message = [@"iBeacon monitoring is supported" mutableCopy];
    }
    
    if(self.logging){
        [self logMessage:message];
    }
    

    
    return enabled;
}

-(BOOL)isEnabled{
    return [CLLocationManager isMonitoringAvailableForClass:[CLRegion class]] &&[CLLocationManager authorizationStatus ]== kCLAuthorizationStatusAuthorizedAlways;
}

-(BOOL)canDeviceSupportAppBackgroundRefresh
{
    // Override point for customization after application launch.
    if ([[UIApplication sharedApplication] backgroundRefreshStatus] == UIBackgroundRefreshStatusAvailable) {
        NSString * message = @"Background updates are available for the app.";
        if(self.logging){
            [self logMessage:message];
        }
        
        
        return YES;
    }else if([[UIApplication sharedApplication] backgroundRefreshStatus] == UIBackgroundRefreshStatusDenied)
    {

        NSString * message = @"The user explicitly disabled background behavior for this app or for the whole system.";
        if(self.logging){
            [self logMessage:message];
        }
        
        return NO;
    }
    else if([[UIApplication sharedApplication] backgroundRefreshStatus] == UIBackgroundRefreshStatusRestricted)
    {
        NSString * message = @"Background updates are unavailable and the user cannot enable them again. For example, this status can occur when parental controls are in effect for the current user.";
        if(self.logging){
            [self logMessage:message];
        }

        return NO;
    }
    return NO;
}
/**Start monitoring regions */
-(void)startMonitoring{
    for (CLBeaconRegion * beaconRegion in self.regions) {
        [self.locationManager startMonitoringForRegion:beaconRegion];
        [self.locationManager startRangingBeaconsInRegion:beaconRegion];
        [self.locationManager startUpdatingLocation];
        [self.locationManager performSelector:@selector(requestStateForRegion:) withObject:beaconRegion afterDelay:1];
    }
}

//Stops Monitoring Services
-(void)stopMonitoring;{
    for (CLBeaconRegion * beaconRegion in self.regions) {
        [self.locationManager stopMonitoringForRegion:beaconRegion];
        [self.locationManager stopRangingBeaconsInRegion:beaconRegion];
        [self.locationManager stopUpdatingLocation];
    }

}

/**
 Register beacons only using identifier and proximity uiid
 */
-(void)registerBeaconWithProximityId:(NSString*)pid andIdentifier:(NSString *)identifier{
    NSUUID *proximityUUID = [[NSUUID alloc]
                             initWithUUIDString:pid];
    
    CLBeaconRegion *beaconRegion = [[CLBeaconRegion alloc]initWithProximityUUID:proximityUUID identifier:identifier];
    
    beaconRegion.notifyOnEntry=YES;
    beaconRegion.notifyOnExit=YES;
    beaconRegion.notifyEntryStateOnDisplay=YES;

    [self.regions addObject:beaconRegion];
    
}

/**
 Estimote beacons use a fixed Proximity UUID of B9407F30-F5F8-466E-AFF9-25556B57FE6D.
 
 Each beacon has a unique ID formatted as follows: proximityUUID.major.minor. We reserved the proximityUUID for all our beacons. The major and minor values are randomized by default but can be customized.
 */

-(void)registerRegionWithProximityId:(NSString*)pid andIdentifier:(NSString *)identifier major:(int)major andMinor:(int)minor{
    NSUUID *proximityUUID = [[NSUUID alloc]
                             initWithUUIDString:pid];
    
    if(major==-1 && minor==-1){
        [self registerBeaconWithProximityId:pid andIdentifier:identifier];
        return;
    }
    
    CLBeaconRegion *beaconRegion;
    beaconRegion= [[CLBeaconRegion alloc]initWithProximityUUID:proximityUUID major:major minor:minor identifier:identifier];
    beaconRegion.notifyOnEntry=YES;
    beaconRegion.notifyOnExit=YES;
    beaconRegion.notifyEntryStateOnDisplay=YES;
    
    [self.regions addObject:beaconRegion];
}

/**Tells the delegate that the user enter  specified region.*/
- (void)locationManager:(CLLocationManager *)manager didEnterRegion:(CLRegion *)region{
    
    if(self.logging){
        NSString * log = [NSString stringWithFormat:@"%s",__PRETTY_FUNCTION__];
        [self logMessage:log];
    }
    
    if([region isKindOfClass:[CLBeaconRegion class]]){
        [self.locationManager startRangingBeaconsInRegion:(CLBeaconRegion *) region];
    }
    
    
    
}


/** Tells the delegate that the user left the specified region.*/
- (void)locationManager:(CLLocationManager *)manager didExitRegion:(CLRegion *)region{
  
    if(self.logging){
        NSString * log = [NSString stringWithFormat:@"%s",__PRETTY_FUNCTION__];
        [self logMessage:log];
    }
    
    proximity = -1;
    
    if([region isKindOfClass:[CLBeaconRegion class]]){
       [self.locationManager stopRangingBeaconsInRegion:(CLBeaconRegion *)region];
       
    }
}

/** Tells the delegate about the state of the specified region. (required) */
- (void)locationManager:(CLLocationManager *)manager didDetermineState:(CLRegionState)state forRegion:(CLRegion *)region{
    NSString * log = [NSString stringWithFormat:@"%s",__PRETTY_FUNCTION__];


    if([region isKindOfClass:[CLBeaconRegion class]]){   //check if the region is beacon region
        [self updateState:(CLBeaconRegion *)region state:state];
  
        if(self.logging){
            [self logMessage:log];
            [self logMessage:[NSString stringWithFormat:@"State for region: %@ is: %d %@ %@",region, (int)state, [(CLBeaconRegion *) region major], [(CLBeaconRegion *) region minor]]];
        }
        
    if(self.regionEvent){
        self.regionEvent( [[(CLBeaconRegion *) region  proximityUUID]UUIDString], [[(CLBeaconRegion *) region major]intValue],[[(CLBeaconRegion *) region minor]intValue],(NSUInteger)state );
    }
  
    if(state == CLRegionStateInside){
             //start ranging beacons
            [self.locationManager startRangingBeaconsInRegion:(CLBeaconRegion *) region];
            [self locationManager:self.locationManager didEnterRegion:region];
        
        }
        if(state == CLRegionStateOutside){
            //start ranging beacons
            [self.locationManager stopRangingBeaconsInRegion:(CLBeaconRegion *) region];
            [self locationManager:self.locationManager didExitRegion:region];
            
        }
        
    }
}




- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error
{
    // Delegate of the location manager, when you have an error
    NSLog(@"didFailWithError: %@", error);
    
  //  UIAlertView *errorAlert = [[UIAlertView alloc]     initWithTitle:NSLocalizedString(@"application_name", nil) message:NSLocalizedString(@"location_error", nil) delegate:nil cancelButtonTitle:NSLocalizedString(@"ok", nil) otherButtonTitles:nil];
    
   // [errorAlert show];
    [self logMessage:error.debugDescription];
    
}

/**Tells the delegate that one or more beacons are in range. */
- (void)locationManager:(CLLocationManager *)manager didRangeBeacons:(NSArray *)beacons inRegion:(CLBeaconRegion *)region{
    
    if(self.nearbyBeacons){
        self.nearbyBeacons(beacons);
    }
    
    for(CLBeacon *beacon in beacons)
    {
            [self checkStateForRegion:region];
            proximity = beacon.proximity;
          //  [self logMessage:[NSString stringWithFormat:@"Beacon range: %@",beacon]];
        
            if(self.beaconFound){
                self.beaconFound(beacon.proximityUUID.UUIDString, beacon.major.intValue, beacon.minor.intValue, beacon.proximity, beacon.accuracy,beacon.rssi);
                if(self.logging){
                    NSString *message = [NSString stringWithFormat:@"Proximity: %ld",  beacon.proximity];
                    [self logMessage:message];
                    //[self log:message];
                
                }
            }
    }
    
    if(counter>30==1){
       
        //[self logMessage:[NSString stringWithFormat:@"%s",__PRETTY_FUNCTION__]];
       counter =0;
    }
    counter++;

}

/** this method can be used to display a content related to the closest beacon */
-(void)displayContentFor:(CLBeacon * )beacon andRegion:(CLRegion *)region{
    if(!_currentBeacon){
        _currentBeacon = beacon;
    }
    else{
        if([_currentBeacon.proximityUUID isEqual:_currentBeacon.proximityUUID]&&_currentBeacon.major ==_currentBeacon.major&&_currentBeacon.minor == _currentBeacon.minor)
        {
            
            if(_currentBeacon.proximity == beacon.proximity){
                //don't change it
                //get content for beacon
                
            } //same beacon but different proximity
            else{
                
            }
        }
    }
}

/** Tells the delegate that an error occurred while gathering ranging information for a set of beacons. */
- (void)locationManager:(CLLocationManager *)manager rangingBeaconsDidFailForRegion:(CLBeaconRegion *)region withError:(NSError *)error{
    NSString * log = [NSString stringWithFormat:@"Failed: %@ %s", error, __PRETTY_FUNCTION__];
    if(self.logging){
    [self logMessage:log];
    }
    
    
}

/** Tells the delegate that a new region is being monitored.*/
- (void)locationManager:(CLLocationManager *)manager didStartMonitoringForRegion:(CLRegion *)region{
    [self.locationManager requestStateForRegion:region];
    NSString * log = [NSString stringWithFormat:@"%s",__PRETTY_FUNCTION__];
    if(self.logging){
        [self logMessage:log];
    }
        //NSLog(@"%@",log);
    // NSLog(@"%@",region);
    
}

/** Tells the delegate that the delivery of location updates has resumed.*/
- (void)locationManagerDidPauseLocationUpdates:(CLLocationManager *)manager{
    NSString * log = [NSString stringWithFormat:@"%s",__PRETTY_FUNCTION__];
    if(self.logging){
        [self logMessage:log];
    NSLog(@"%@ %s",log, __PRETTY_FUNCTION__);
    }


    
}
/**
 *  Get content of a log file as a string
 *
 *  @return string with log information
 */

-(NSString *)getLog{
    NSString * log = @"";
    NSString * docs = [self applicationDocumentsDirectory];
    NSString * filePath = [docs stringByAppendingPathComponent:@"log.txt"];
    if([[NSFileManager defaultManager]fileExistsAtPath:filePath]){
        NSError *e;
        NSString * existingFile = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:&e];
        if(e){
            NSLog(@"%s %@",__PRETTY_FUNCTION__, e.debugDescription);
        }
        else{
            log = existingFile;
        }
    }
    return log;
}

/**
 *  Save's a log to txt log file
 *
 *  @param string Message to save
 */
-(void)saveLog:(NSString *)string{
    
    NSString * stringToSave= @"";
    NSString * docs = [self applicationDocumentsDirectory];
    NSString * filePath = [docs stringByAppendingPathComponent:@"log.txt"];
    if([[NSFileManager defaultManager]fileExistsAtPath:filePath]){
        NSError *e;
      //  NSData * d= [NSData dataWithContentsOfFile:filePath];
        NSString * existingFile = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:&e];
        if(e){
            NSLog(@"%s %@",__PRETTY_FUNCTION__, e.debugDescription);
        }
        else{
            stringToSave = [NSString stringWithFormat:@"/r/n %@ %@", existingFile, string];
            
        }
    }
    else{
        stringToSave = string;
    }
   NSData *d = [stringToSave dataUsingEncoding:NSUTF8StringEncoding];
    
    [d writeToFile:filePath atomically:YES];
    
}
/**
 *  Immortal documents path
 *
 *  @return document's path
 */
- (NSString *) applicationDocumentsDirectory
{
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *basePath = ([paths count] > 0) ? [paths objectAtIndex:0] : nil;
    return basePath;
}



@end

File: JMCiBeaconManager/Classes/JMCBeaconManager.swift
//
//  JMCBeaconManager.swift
//  PerfecTour
//
//  Created by Janusz Chudzynski on 1/30/16.
//  Copyright © 2016 PerfecTour. All rights reserved.

import UIKit
import CoreLocation



/**Used to broadcast NSNotification*/
public enum iBeaconNotifications:String{
    case BeaconProximity
    case BeaconState
    case Location // new location discoverd
    case iBeaconEnabled
    case iBeaconDisabled

}

/**Interacting with the iBeacons*/
public class JMCBeaconManager: NSObject, CLLocationManagerDelegate {

    let locationManager:CLLocationManager = CLLocationManager()
//    private var beacons = [iBeacon]() // Currently unused
    /**Storing reference to registered regions*/
    private var regions = [CLBeaconRegion]()
    
    var stateCallback:((beacon:iBeacon)->Void)?
    var rangeCallback:((beacon:iBeacon)->Void)?
    var bluetoothManager:BluetoothManager?
    
    var logging = true
    var broadcasting = true
   
    //Different cases
    var bluetoothDisabled = true
 
    /**Error Callback*/
    var errorCallback:((messages:[String])->Void)?
    
    /**Success Callback*/
    var successCallback:(()->Void)?
    
    
    override init(){
    
        super.init()
//        bluetoothManager.callback = bluetoothUpdate
        locationManager.delegate = self
        registerNotifications()
        //locationManager.startUpdatingLocation()
        //test if enabled
    }
    

   
    
    

    /**Starts Monitoring for beacons*/
    func startMonitoring(successCallback:(()->Void), errorCallback:(messages:[String])->Void){
        self.successCallback = successCallback
        self.errorCallback = errorCallback
         checkStatus()
    }
    
    /**Checks the status of the application*/
    func checkStatus(){
        //starts from Bluetooth
        if let _ = self.bluetoothManager{
        
        }
        else{
            bluetoothManager = BluetoothManager()
            bluetoothManager?.callback = bluetoothUpdate
        }
        
    }
    
    
   // var bluetoothManager
    
    /**Check Bluetooth*/
     private func bluetoothUpdate(status:Bool)->Void{
        if status == true{
             bluetoothDisabled = false
            //rund additional status check
            let tuple = statusCheck()
            if tuple.0{
               self.successCallback?()
               NSNotificationCenter.defaultCenter().postNotificationName(iBeaconNotifications.iBeaconEnabled.rawValue, object: nil)
               startMonitoring()

            }
            else{
                self.errorCallback?(messages: tuple.messages)
                NSNotificationCenter.defaultCenter().postNotificationName(iBeaconNotifications.iBeaconDisabled.rawValue, object: tuple.messages)
            }
        }
        else{
            NSNotificationCenter.defaultCenter().postNotificationName(iBeaconNotifications.iBeaconDisabled.rawValue, object:nil)
            self.errorCallback?(messages: ["Bluetooth not enabled."])
            bluetoothDisabled = true
        }
    }
    
    /**Checks if ibeacons are enabled. Should be called first*/
    private func statusCheck()->(Bool,messages:[String]){
        
        locationManager.requestAlwaysAuthorization()
        var check = true
        var messages = [String]()
        
        if bluetoothDisabled == true
        {
            messages.append("Bluetooth must be turned on.")
            check = false
        }
        
        
        if CLLocationManager.authorizationStatus() != .AuthorizedAlways {
            if logging {
                print("Error - authorization status not enabled!")
            }
            messages.append("Location Services Must be Authorized.")
            check = false
        }
        
        if !CLLocationManager.isMonitoringAvailableForClass(CLRegion){
            check = false
            messages.append("CLLocationManager monitoring is not enabled on this device.")

        }
    
        return (check, messages)
    }
    //
    
    /**Register Notifications*/
    func registerNotifications()
    {
        NSNotificationCenter.defaultCenter().addObserverForName(UIApplicationBackgroundRefreshStatusDidChangeNotification, object: UIApplication.sharedApplication(), queue: nil) { (notification) -> Void in
            
        }
        
        
    }
    
    /**Register iBeacons*/
    func registerBeacons(beacons:[iBeacon])
    {
        
        for beacon in beacons{
            
           var beaconRegion:CLBeaconRegion = CLBeaconRegion(proximityUUID: NSUUID(UUIDString:beacon.UUID)!, identifier: beacon.id)
            
            /**Only major infoermation provided*/
            if let major = beacon.major{
                beaconRegion = CLBeaconRegion(proximityUUID: NSUUID(UUIDString:beacon.UUID)!, major: major, identifier: beacon.id)
            }
            
            /**All the information provided*/
            if let major = beacon.major, minor = beacon.minor{
                beaconRegion = CLBeaconRegion(proximityUUID: NSUUID(UUIDString:beacon.UUID)!, major: major, minor: minor, identifier: beacon.id)
            }
  
            beaconRegion.notifyEntryStateOnDisplay = true
            beaconRegion.notifyOnEntry = true
            beaconRegion.notifyOnExit = true

            regions.append( beaconRegion)
         }
    }

    
    /**Register iBeacons*/
    func registerBeacon(beaconId:String)
    {
        
        let bid = CLBeaconRegion(proximityUUID:  NSUUID(UUIDString:beaconId)!, identifier: "Testing Beacon")
        regions.append(bid)
    }

    
    
    /**Starts monitoring beacons*/
    func startMonitoring(){
        locationManager.startUpdatingLocation()
        locationManager.distanceFilter  = 10
        locationManager.desiredAccuracy = kCLLocationAccuracyNearestTenMeters

        for beaconRegion in regions{
            locationManager.startMonitoringForRegion(beaconRegion)
            locationManager.startRangingBeaconsInRegion(beaconRegion)
            //FIXME: check if needed [self.locationManager performSelector:@selector(requestStateForRegion:) withObject:beaconRegion afterDelay:1];
            //FIXME: added more validation for the ibeacons permission matrix
        }
    
    }

    /**Starts monitoring beacons*/
    func stopMonitoring(){
        for beaconRegion in regions{
            locationManager.stopMonitoringForRegion(beaconRegion)
            locationManager.stopRangingBeaconsInRegion(beaconRegion)          
        }
          locationManager.stopUpdatingLocation()
    }

    
    // MARK: Core Location Delegate
    /*
    *  locationManager:didDetermineState:forRegion:
    *
    *  Discussion:
    *    Invoked when there's a state transition for a monitored region or in response to a request for state via a
    *    a call to requestStateForRegion:.
    */
    
    func locationManager(manager: CLLocationManager, didDetermineState state: CLRegionState, forRegion region: CLRegion)
    {
        
        //we found a beacon. Now
        if let region = region as? CLBeaconRegion
        {
            if logging {
                print("State determined\(region) \(state.rawValue)")
            }
        }
        
        //we found a beacon. Now
        if let region = region as? CLBeaconRegion, minor = region.minor, major  = region.major{
            
            let beacon = iBeacon(minor: minor.unsignedShortValue, major: major.unsignedShortValue, proximityId: region.proximityUUID.UUIDString)
            beacon.state = state
            
            if logging {
                print("State determined\(region) \(state.rawValue)")
            }
    
            if broadcasting{
                //broadcast notification
                //get beacon from clregion
 
                NSNotificationCenter.defaultCenter().postNotificationName(iBeaconNotifications.BeaconState.rawValue, object: beacon)
            }
            
            if let callback = self.stateCallback{
                callback(beacon: beacon)
            }
        }
        //if we are outside stop ranging
        if state == .Outside{
            manager.stopRangingBeaconsInRegion(region as! CLBeaconRegion)
        }
        if state == .Inside{
            manager.startRangingBeaconsInRegion(region as! CLBeaconRegion)
        }
        
        
    }
    /*
    *  locationManager:didRangeBeacons:inRegion:
    *
    *  Discussion:
    *    Invoked when a new set of beacons are available in the specified region.
    *    beacons is an array of CLBeacon objects.
    *    If beacons is empty, it may be assumed no beacons that match the specified region are nearby.
    *    Similarly if a specific beacon no longer appears in beacons, it may be assumed the beacon is no longer received
    *    by the device.
    */
    
    func locationManager(manager: CLLocationManager, didRangeBeacons beacons: [CLBeacon], inRegion region: CLBeaconRegion)
    {
        //Notify the delegates and etc that we know how far are we from the iBeacon
        if logging {
            print("Did Range Beacons \(beacons)")
        }
        
        if broadcasting{
        //broadcast notification
        //convert CLBeacon to iBeacon 
        var myBeacons = [iBeacon]()
            //convert it to our iBeacons
            for beacon in beacons{
                if beacon.proximity != .Unknown{
                    let myBeacon = iBeacon(beacon: beacon)
                    myBeacons.append(myBeacon)
                }
            }

        myBeacons.sortInPlace({$0.proximity.sortIndex < $1.proximity.sortIndex})
            
            NSNotificationCenter.defaultCenter().postNotificationName(iBeaconNotifications.BeaconProximity.rawValue, object: myBeacons)
        }

        
        for beacon in beacons{
            if logging {
                print("Did Range Beacon \(beacon)")
            }
            if let callback = self.rangeCallback{
                //convert it to the internal type of the beacon
               let ibeacon =  iBeacon(minor: beacon.minor.unsignedShortValue, major: beacon.major.unsignedShortValue, proximityId: beacon.proximityUUID.UUIDString)
                ibeacon.proximity = beacon.proximity
                callback(beacon: ibeacon)
            }
        }
    }
    
    
    /**Update Location*/
    func locationManager(manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        
        if let location = locations.last{
            print("Update Location to \(location)")
            NSNotificationCenter.defaultCenter().postNotificationName(iBeaconNotifications.Location.rawValue,  object: location)
        }
    }
    
    /*
    *  locationManager:rangingBeaconsDidFailForRegion:withError:
    *
    *  Discussion:
    *    Invoked when an error has occurred ranging beacons in a region. Error types are defined in "CLError.h".
    */

    func locationManager(manager: CLLocationManager, rangingBeaconsDidFailForRegion region: CLBeaconRegion, withError error: NSError)
    {
        if logging {
            print("Ranging Fail\(region) \(error.debugDescription)")
        }
    }

    
    /*
    *  locationManager:didEnterRegion:
    *
    *  Discussion:
    *    Invoked when the user enters a monitored region.  This callback will be invoked for every allocated
    *    CLLocationManager instance with a non-nil delegate that implements this method.
    */
    func locationManager(manager: CLLocationManager, didEnterRegion region: CLRegion)
    {
        if region is CLBeaconRegion{
            if logging {
                print("Region Entered! \(region) ")
                manager.startRangingBeaconsInRegion(region as! CLBeaconRegion)
            }
        }
    }
    /*
    *  locationManager:didExitRegion:
    *
    *  Discussion:
    *    Invoked when the user exits a monitored region.  This callback will be invoked for every allocated
    *    CLLocationManager instance with a non-nil delegate that implements this method.
    */

    func locationManager(manager: CLLocationManager, didExitRegion region: CLRegion)
    {
        if region is CLBeaconRegion{
            if logging {
                print("Exit Region! \(region) ")
                manager.stopRangingBeaconsInRegion(region as! CLBeaconRegion)
            }
        }
    }
    /*
    *  locationManager:didFailWithError:
    *
    *  Discussion:
    *    Invoked when an error has occurred. Error types are defined in "CLError.h".
    */
    
    func locationManager(manager: CLLocationManager, didFailWithError error: NSError)
    {
       
            if logging {
                print("Manager Failed with error \(error)")
            }

    }
    /*
    *  locationManager:monitoringDidFailForRegion:withError:
    *
    *  Discussion:
    *    Invoked when a region monitoring error has occurred. Error types are defined in "CLError.h".
    */

    func locationManager(manager: CLLocationManager, monitoringDidFailForRegion region: CLRegion?, withError error: NSError)
    {
        if logging {
            print("Monitoring Failed with error \(error)")
        }

    }
    /*
    *  locationManager:didChangeAuthorizationStatus:
    *
    *  Discussion:
    *    Invoked when the authorization status changes for this application.
    */

    func locationManager(manager: CLLocationManager, didChangeAuthorizationStatus status: CLAuthorizationStatus)
    {
        if status == CLAuthorizationStatus.AuthorizedAlways{
            if statusCheck().0 == true {
                startMonitoring()
            }
        }
    }
    /*
    *  locationManager:didStartMonitoringForRegion:
    *
    *  Discussion:
    *    Invoked when a monitoring for a region started successfully.
    */
    
    func locationManager(manager: CLLocationManager, didStartMonitoringForRegion region: CLRegion)
    {
    
    }
    /*
    *  Discussion:
    *    Invoked when location updates are automatically paused.
    */

    func locationManagerDidPauseLocationUpdates(manager: CLLocationManager)
    {
    
    }
    /*
    *  Discussion:
    *    Invoked when location updates are automatically resumed.
    *
    *    In the event that your application is terminated while suspended, you will
    *	  not receive this notification.
    */
    
    func locationManagerDidResumeLocationUpdates(manager: CLLocationManager)
    {
    
    } 
}
